### PR TITLE
type hinting for hashes, hmac, and cmac

### PR DESCRIPTION
**Pull request by reaperhulk**

### CODE REVIEW

File: src/cryptography/hazmat/backends/openssl/cmac.py [hunk 1]
```diff
@@ -65,7 +65,7 @@ def finalize(self):
 
         return self._backend._ffi.buffer(buf)[:]
 
-    def copy(self):
+    def copy(self) -> "_CMACContext":
```
Comment (alex) [inline src/cryptography/hazmat/backends/openssl/cmac.py:68]: Should this be the interface, not the concrete type?

File: src/cryptography/hazmat/backends/openssl/hashes.py [hunk 2]
```diff
@@ -37,7 +36,7 @@ def __init__(self, backend, algorithm, ctx=None):
 
     algorithm = utils.read_only_property("_algorithm")
 
-    def copy(self):
+    def copy(self) -> "_HashContext":
```
Comment (alex) [inline src/cryptography/hazmat/backends/openssl/hashes.py:39]: same.

File: src/cryptography/hazmat/backends/openssl/hmac.py [hunk 3]
```diff
@@ -41,7 +42,7 @@ def __init__(self, backend, key, algorithm, ctx=None):
 
     algorithm = utils.read_only_property("_algorithm")
 
-    def copy(self):
+    def copy(self) -> "_HMACContext":
```
Comment (alex) [inline src/cryptography/hazmat/backends/openssl/hmac.py:45]: Same 😂

File: src/cryptography/hazmat/backends/openssl/cmac.py (hunk 1 above)
Comment (reaperhulk) [reply src/cryptography/hazmat/backends/openssl/cmac.py:68]: There's no CMAC interface 😄. For HMAC and Hash we also have a structure where we instantiate an object which in turn instantiates a context. Every user-facing operation goes through the primary object, which in turn invokes the underlying context. So the user will never actually interact with these context objects directly.

**Review by alex — APPROVED** [review]

**Pull request merged**